Tests/HummingbirdRouterTests/ControllerTests.swift

```diff
@@ -0,0 +1,78 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Hummingbird server framework project
+//
+// Copyright (c) 2021-2023 the Hummingbird authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See hummingbird/CONTRIBUTORS.txt for the list of Hummingbird authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import Hummingbird
+import HummingbirdRouter
+import HummingbirdTesting
+import XCTest
+
+final class ControllerTests: XCTestCase {
+    func testRouterController() async throws {
```

Comment:
We should add tests that test all of the functions in `MiddlewareFixedTypeBuilder` so we need a test with a single route inside a controller.

Reply:
I added a few tests, and actually removed some of the added result builders.

I think we don't actually need any of the additional result builders except for `buildExpression` Looking at the code coverage, basically all the implementations flow through the `RouterController` `buildExpression` and then back into the normal result builders defined in `Hummingbird`

```
public static func buildExpression<C0: RouterController>(_ c0: C0) -> C0.Body where C0.Body.Input == Input, C0.Body.Output == Output, C0.Body.Context == Context {
    return c0.body
}
```

Reply:
I wish that parameter packs supported Same-element requirements, it would allow us to replace `_Middleware2` with something like this

```
public struct TupleMiddleware<Input, Output, Context, each M: Handler<Input, Output, Context>>: MiddlewareProtocol {
    public func handle(_ input: Input, context: Context) -> Output {
        ...
    }
}
```

this would greatly reduce the type signature craziness when looking at errors.

i.e stuff like this 

`Generic struct 'ParentController' requires that '_Middleware2<_Middleware2<_Middleware2<Route<_RouteHandlerClosure<String, BasicRouterRequestContext>, BasicRouterRequestContext>, some RouterMiddleware<BasicRouterRequestContext>>, some RouterMiddleware<BasicRouterRequestContext>>, Route<_RouteHandlerClosure<String, BasicRouterRequestContext>, BasicRouterRequestContext>>' ....etc......`

Reply:
I'm assuming you'd be opposed to the old-fashioned pre-parameter-pack `_Middleware2`, `_Middleware3`, `_Middleware4` etc...

Reply:
I'm not going to change it now. I can't see how you'd do it without breaking changes. 

The MiddlewareProtocol is closely mapped to a generic middleware being developed by members of SSWG (unavailable at the moment). It's just taking a long time because key members are very busy.

Reply:
> I think we don't actually need any of the additional result builders except for buildExpression

Ok if you can show via tests that all expected cases are covered by `buildExpression` then we should remove the others.

Reply:
I think the additional tests should cover the basis. I'll do another pass in an hour or so, just popping out for an early dinner.

Reply:
I added one more test to validate middleware compiles correctly and executes.